src/adapter/src/catalog/open/builtin_item_migration.rs

```diff
@@ -64,27 +76,186 @@ pub(crate) async fn migrate_builtin_items(
         read_only,
     }: BuiltinItemMigrationConfig,
 ) -> Result<BuiltinItemMigrationResult, Error> {
-    migrate_builtin_items_0dt(
+    assert_eq!(
+        read_only,
+        txn.is_savepoint(),
+        "txn must be in savepoint mode when read_only is true, and in writable mode otherwise",
+    );
+
+    update_catalog_fingerprints(state, txn, &migrated_builtins)?;
+
+    // Collect GlobalIds of storage collections we need to migrate.
+    let mut collections_to_migrate: Vec<_> = migrated_builtins
+        .into_iter()
+        .filter_map(|id| {
+            use CatalogItem::*;
+            match &state.get_entry(&id).item() {
+                Table(table) => Some(table.global_ids().into_element()),
+                Source(source) => Some(source.global_id()),
+                MaterializedView(mv) => Some(mv.global_id()),
+                ContinualTask(ct) => Some(ct.global_id()),
+                Log(_) | Sink(_) | View(_) | Index(_) | Type(_) | Func(_) | Secret(_)
+                | Connection(_) => None,
+            }
+        })
+        .collect();
+
+    // Attempt to perform schema evolution.
+    //
+    // If we run into an unexpected error while trying to perform schema evolution, we abort the
```

Comment:
Can we unpack what an expected/unexpected error is here? (IIUC a schema incompat error is "expected" but other errors are not.)